src/main/java/org/opensearch/geospatial/ip2geo/action/IpEnrichmentTransportAction.java

```diff
@@ -0,0 +1,87 @@
+import org.opensearch.geospatial.ip2geo.dao.DatasourceDao;
+import org.opensearch.geospatial.ip2geo.dao.Ip2GeoCachedDao;
+import org.opensearch.geospatial.ip2geo.jobscheduler.Datasource;
+import org.opensearch.tasks.Task;
+import org.opensearch.transport.TransportService;
+
+import lombok.extern.log4j.Log4j2;
+
+/**
+ * Transport action to convert provided IP address String into GeoLocation data.
+ */
+@Log4j2
+public class IpEnrichmentTransportAction extends HandledTransportAction<ActionRequest, ActionResponse> {
+
+    private final Ip2GeoCachedDao ip2GeoCachedDao;
+
+    private final DatasourceDao datasourceDao;
+
+    /**
+     * Constructor
+     * @param transportService the transport service
+     * @param actionFilters the action filters
+     * @param cachedDao the cached datasource facade
+     * @param datasourceDao the datasourceDao to read the default datasource from
+     */
+    @Inject
+    public IpEnrichmentTransportAction(
+        TransportService transportService,
+        ActionFilters actionFilters,
+        Ip2GeoCachedDao cachedDao,
+        DatasourceDao datasourceDao
+    ) {
+        super(IpEnrichmentAction.NAME, transportService, actionFilters, IpEnrichmentRequest::new);
+        this.ip2GeoCachedDao = cachedDao;
+        this.datasourceDao = datasourceDao;
+    }
+
+    /**
+     * Overridden method to extract IP String from IpEnrichmentRequest object and return the enrichment result
+     * in the form of IpEnrichmentResponse which contains the GeoLocation data for given IP String.
+     * @param task the task.
+     * @param request request object in the form of IpEnrichmentRequest which contain the IP String to resolve
+     * @param listener a container which encapsulate IpEnrichmentResponse object with the GeoLocation data for given IP.
+     */
+    @Override
+    protected void doExecute(Task task, ActionRequest request, ActionListener<ActionResponse> listener) {
+        IpEnrichmentRequest enrichmentRequest = IpEnrichmentRequest.fromActionRequest(request);
+        String ipString = enrichmentRequest.getIpString();
+        String defaultDataSourceName = getDefaultDataSourceName();
+        if (enrichmentRequest.getDatasourceName() == null && defaultDataSourceName == null) {
+            log.error("No data source available, IpEnrichmentTransportAction aborted.");
+            listener.onFailure(new IllegalArgumentException());
+        } else {
+            String dataSourceName = Optional.ofNullable(enrichmentRequest.getDatasourceName()).orElse(defaultDataSourceName);
+            String indexName = ip2GeoCachedDao.getIndexName(dataSourceName);
+            Map<String, Object> geoLocationData = ip2GeoCachedDao.getGeoData(indexName, ipString);
+            log.debug("GeoSpatial IP lookup on IP: [{}], and result [{}]", ipString, geoLocationData);
+            listener.onResponse(new IpEnrichmentResponse(geoLocationData));
+        }
+    }
+
+    private String getDefaultDataSourceName() {
```

Comment:
Would it be better to make the data source name mandatory? Allowing the caller to omit the data source and selecting one randomly from the available options might not be ideal.